bitcoin/BRSyncManager.c

```diff
@@ -1232,62 +1221,178 @@ BRClientSyncManagerGenerateRid (BRClientSyncManager manager) {
+        scanState->lastExternalAddress = externalAddress;
+        scanState->lastInternalAddress = internalAddress;
+
+         newAddresses = _updateWalletAddressSet (scanState->knownAddresses, wallet);
+    }
+
+    return newAddresses;
 }
 
 static BRAddress *
-BRClientSyncManagerGetAllAddrs (BRClientSyncManager manager,
-                                size_t *addressCount) {
+_getWalletAddresses (BRWallet *wallet,size_t *addressCount) {
     assert (addressCount);
 
-    size_t addrCount = BRWalletAllAddrs (manager->wallet, NULL, 0);
+    size_t addrCount = BRWalletAllAddrs (wallet, NULL, 0);
 
     BRAddress *addrs = (BRAddress *) calloc (2 * addrCount, sizeof (BRAddress));
-    BRWalletAllAddrs (manager->wallet, addrs, addrCount);
+    BRWalletAllAddrs (wallet, addrs, addrCount);
 
     memcpy (addrs + addrCount, addrs, addrCount * sizeof(BRAddress));
     for (size_t index = 0; index < addrCount; index++)
-        BRClientSyncManagerAddressToLegacy (manager, &addrs[addrCount + index]);
+        addrs[addrCount + index] = BRWalletAddressToLegacy (wallet, &addrs[index]);
 
     *addressCount = 2 * addrCount;
     return addrs;
 }
 
-/**
- * Return all addresses, used and unused, tracked by the wallet. The addresses
- * are both 'internal' and 'external' ones.
- *
- * The addresses are returned as both a sequential array of BRAddress data, as well
- * as an array of pointers to each address.
- *
- * Note: Both the addressStrings and addressArray arrays must be freed.
- *
- * Note: The addressStrings array contains pointers to data in the addressArray. As such,
- *       elements in addressStrings should not be accessed once addressArray has been freed.
- */
 static void
-BRClientSyncManagerGetAllAddrsAsStrings (BRClientSyncManager manager,
-                                         size_t *addressCount,
-                                         const char ***addressStrings,
-                                         BRAddress **addressArray) {
-    size_t addrCount     = 0;
-    BRAddress *addrArray = BRClientSyncManagerGetAllAddrs (manager, &addrCount);
-
-    const char **addrsStrings = calloc (addrCount, sizeof(char *));
-    for (size_t index = 0; index < addrCount; index ++)
-        addrsStrings[index] = (char *) &addrArray[index];
-
-    *addressCount = addrCount;
-    *addressStrings = addrsStrings;
-    *addressArray = addrArray;
+_fillWalletAddressSet(BRSetOf(BRAddress *) addresses, BRWallet *wallet) {
+    size_t addressCount = 0;
+    BRAddress *addressArray = _getWalletAddresses (wallet, &addressCount);
+
+    for (size_t index = 0; index < addressCount; index++) {
+        if (!BRSetContains (addresses, &addressArray[index])) {
+            BRAddress *address = malloc (sizeof(BRAddress));
+            *address = addressArray[index];
+            BRSetAdd (addresses, address);
+        }
+    }
+
+    free (addressArray);
+}
+
+static BRArrayOf(char *)
+_updateWalletAddressSet(BRSetOf(BRAddress *) addresses, BRWallet *wallet) {
```

Comment:
This could return BRArrayOf(BRAddress) - the array_* macros will recognize BRAddress as a struct and allocate those chunks.  Doing so avoid the subsequence `malloc` in the loop.  I think.

I think a similar issue above in _filWalletAddressSet() - don't need to use `BRAddress *` and thus the `malloc()` (and subsequent `free()`

Reply:
I need an array of string pointers to pass to the BTC getTransactions call, not an array of BRAddress structs. Also, the BRSet holders pointers to items.

I also hoped to avoid having to do this but did not see a path (short of changing the getTransactions callback to expect BRAddress[], which I didn't think was a good idea).

Ultimately, this is C and the performance impact of allocations here isn't really an issue, I don't believe.

Reply:
That is fine.  There are or can be BRSetOf 'BRAddress', 'BRAddress*' and 'char *'.  I was suggesting using 'BRAddress' for BRSet and then create BRArrayOf 'char *' from addr.s.  But, either is fine.